src/OpenSage.Game/Logic/Object/ExperienceTracker.cs

```diff
@@ -0,0 +1,257 @@
+            return;
+        }
+
+        var oldLevel = _currentLevel;
+        var amountToGain = experienceGain;
+        if (canScaleForBonus)
+        {
+            amountToGain = (int)(amountToGain * ExperienceScalar);
+        }
+
+        _currentExperience += amountToGain;
+
+        var levelIndex = 0;
+
+        // While we can gain more levels and we have enough experience to gain the next level...
+        while (((levelIndex + 1) < (int)VeterancyLevel.Count) && _currentExperience >= _parent.Definition.ExperienceRequired[(VeterancyLevel)(levelIndex + 1)])
+        {
+            levelIndex++;
+        }
+
+        _currentLevel = (VeterancyLevel)levelIndex;
+
+        // If this experience gain resulted in a level up, notify the parent object.
+        if (oldLevel != _currentLevel)
+        {
+            OnVeterancyLevelChanged(oldLevel, _currentLevel);
+        }
+    }
+
+    public bool GainExpForLevel(int levelsToGain, bool canScaleForBonus = true)
+    {
+        var newLevel = _currentLevel + levelsToGain;
+
+        if (newLevel > VeterancyLevel.Last)
+        {
+            newLevel = VeterancyLevel.Last;
+        }
+
+        if (newLevel > _currentLevel)
+        {
+            var experienceNeeded = _parent.Definition.ExperienceRequired[newLevel] - _currentExperience;
+            AddExperiencePoints(experienceNeeded, canScaleForBonus);
+            return true;
+        }
+
+        return false;
+    }
+
+    public bool CanGainExpForLevel(int levelsToGain)
+    {
+        var newLevel = _currentLevel + levelsToGain;
+
+        if (newLevel > VeterancyLevel.Last)
+        {
+            newLevel = VeterancyLevel.Last;
+        }
+
+        return newLevel > _currentLevel;
+    }
+
+    // TODO: This is like 95% identical to AddExperiencePoints.
+    public void SetExperienceAndLevel(int experienceIn, bool provideFeedback = true)
+    {
+        if (ExperienceSink.IsValid)
+        {
+            var sinkObject = _engine.Scene3D.GameObjects.GetObjectById(ExperienceSink);
+
+            if (sinkObject != null)
+            {
+                sinkObject.ExperienceTracker.SetExperienceAndLevel(experienceIn, provideFeedback);
+                return;
+            }
+        }
+
+        if (!IsTrainable)
+        {
+            return;
+        }
+
+        var oldLevel = _currentLevel;
+        _currentExperience = experienceIn;
+
+        var levelIndex = 0;
+        while (((levelIndex + 1) < (int)VeterancyLevel.Count) && _currentExperience >= _parent.Definition.ExperienceRequired[(VeterancyLevel)(levelIndex + 1)])
+        {
+            levelIndex++;
+        }
+
+        _currentLevel = (VeterancyLevel)levelIndex;
+
+        if (oldLevel != _currentLevel)
+        {
+            OnVeterancyLevelChanged(oldLevel, _currentLevel, provideFeedback);
+            // Comment from original code: "paradox! this may be a level lost!"
+            // What does that mean?
```

Comment:
potentially this could be used to set a unit's experience to a lower amount than they currently have?

Reply:
True, though the same applies to `SetVeterancyLevel` as well.

Reply:
I don't think this needs to be solved now or anything, but it's probably worth revisiting at some point. I'm assuming mission scripts can set the veterancy level of a unit, even if it means lowering its rank.